### PR TITLE
Implement car

### PR DESCRIPTION
**Pull request by mladjo2505**
https://trello.com/c/mW5PAJHu/493-lodo-car-md-car

https://trello.com/c/CTQdeGGa/274-lodo-car-department-selected-in-fb-should-be-automatically-downloaded-to-lodo

### CODE REVIEW

File: db/changes/116_add_car_table.sql [hunk 1]
```diff
@@ -0,0 +1,48 @@
+-- Create new tables for Car and Cars milage
+CREATE TABLE IF NOT EXISTS companycar (
+CompanyCarID int(11) NOT NULL AUTO_INCREMENT,
```
Comment (sls-lnluksa) [inline db/changes/116_add_car_table.sql]: Rename CompanyCar into Car. (Table name and id fields)

File: db/changes/116_add_car_table.sql [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+TS timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+Active tinyint(4) NOT NULL DEFAULT '0',
+ValidFrom datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
+ValidTo datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
+Description text,
+CarCode varchar(10) DEFAULT NULL,
+PurchasePrice decimal(16, 2) DEFAULT '0',
+PurchaseAddCost int(11) DEFAULT '0',
+PurchaseYear smallint(6) NOT NULL DEFAULT '0',
+CurrentPrice int(11) NOT NULL DEFAULT '0',
```
Comment (sls-lnluksa) [inline db/changes/116_add_car_table.sql]: Remove
+PurchaseAddCost int(11) DEFAULT '0',
+PurchaseYear smallint(6) NOT NULL DEFAULT '0',
+CurrentPrice int(11) NOT NULL DEFAULT '0',

File: db/changes/116_add_car_table.sql [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+EnableVAT smallint(6) DEFAULT NULL,
+VehicleType varchar(50) DEFAULT '',
+PRIMARY KEY (CompanyCarID)
+);
+
+CREATE TABLE IF NOT EXISTS companycarmilage (
+CompanyCarID int(11) NOT NULL,
+MilageYear smallint(6) NOT NULL DEFAULT '0',
+km0101 int(11) DEFAULT '0',
+km3112 int(11) DEFAULT '0',
```
Comment (sls-lnluksa) [inline db/changes/116_add_car_table.sql]: Rename km0101 to StartMilage
Rename km3112 to EndMilage

**Comment by sls-lnluksa** [comment]
:exclamation: 
- [x] Checkout why CarID is not saved in vouchers
- [x] Checkout why CarID is not saved/displayed in invoicein
- [x] bank.tabjournal change colspan

File: modules/car/view/record.inc [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+elseif($_lib['input']->getProperty('action_car_delete')) {
+  #Possible to extend or alter parameters here
+  $_lib['db']->db_delete_hash($db_table, $primarykey);
+  # delete all milage for the car
+  $query  = "DELETE FROM companycarmilage WHERE CompanyCarID=" . (int) $CompanyCarID . "";
+  $result = $_lib['db']->db_query($query);
+}
+
+if (!$_lib['input']->getProperty('action_car_delete') && $CompanyCarID) {
```
Comment (sls-lnluksa) [inline modules/car/view/record.inc]: Comment why this is here. 
If you change purchase year to newer year, implement logic around it so we delete all mileage records for previous years.

File: modules/car/view/edit.php [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+$car_milage_query = "SELECT * FROM companycarmilage WHERE CompanyCarID = $CompanyCarID";
+$car_milage_result = $_lib['storage']->db_query($car_milage_query);
+while($_car_milage = $_lib['storage']->db_fetch_object($car_milage_result)) {
+  $car_milage[$_car_milage->MilageYear] = $_car_milage;
+}
+
+#Do car calculations
+$car_calculations = array();
+foreach($car_milage as $milage_year => $milage) {
+  if($milage->km0101) {
```
Comment (sls-lnluksa) [inline modules/car/view/edit.php]: Include this logic for StartMilage = 0

File: modules/car/view/edit.php [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  <td class="menu">
+    <table class="menu lodo_data">
+      <tr><td class="menu">Calcualtions</td></tr>
+      <tr><td class="menu">Start milage</td></tr>
+      <tr><td class="menu">End milage</td></tr>
+      <tr><td class="menu">Distance</td></tr>
+      <tr><td class="menu">Money spent on fuel</td></tr>
+      <tr><td class="menu">Money spent per mile</td></tr>
+    </table>
+  </td>
```
Comment (sls-lnluksa) [inline modules/car/view/edit.php:151]: @stoivo Translate this.

File: modules/car/view/edit.php [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        <?
+          $_years = array_keys($car_calculations);
+          rsort($_years);
+          foreach($_years as $year) { ?>
+          <th><? print $year; ?></th>
+        <? } ?>
+      </tr>
+      <tr>
+        <? foreach($_years as $year) { ?>
+          <td><input type="text" name="companycarmilage.km0101.<? print $year; ?>" value="<? print $car_calculations[$year]['km0101']; ?>" size="15"></td>
```
Comment (sls-lnluksa) [inline modules/car/view/edit.php]: align all fields to the right

File: modules/car/view/edit.php [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+<tr>
+    <td class="menu">Kj&oslash;psdato</td>
+    <td><input type="text" name="companycar.ValidFrom" value="<? if ((int)($companycar->ValidFrom) != 0) print strftime("%F", strtotime($companycar->ValidFrom)) ?>" size="60"></td>
+</tr>
+<tr>
+    <td class="menu">Kj&oslash;pepris</td>
+    <td><input type="text" name="companycar.PurchasePrice" value="<? print $_lib['format']->Amount($companycar->PurchasePrice) ?>" size="60"></td>
+</tr>
+<tr>
+    <td class="menu">Kj&oslash;psdato</td>
```
Comment (sls-lnluksa) [inline modules/car/view/edit.php]: Kj&oslash;psdato => Salgsdato

File: modules/car/view/list.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        <? print $_lib['form3']->text(array('table'=>$db_table, 'field'=>'CompanyCarID', 'width'=>'10')) ?>
+        <? print $_lib['form3']->submit(array('name'=>'action_car_new', 'value'=>'Ny bil')) ?>
+      </form>
+    <? } ?>
+    </th>
+  </tr>
+  <tr>
+    <td class="menu">Bil</td>
+    <td class="menu">Bil navn</td>
+    <td class="menu">Aktiv</td>
```
Comment (sls-lnluksa) [inline modules/car/view/list.php:48]: Add CarCode, PurchaseDate, SaleDate

File: modules/car/view/record.inc [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  for($milage_year = $car_purchase_year; $milage_year <= $_year; $milage_year++) {
+    $post_values = array("companycarmilage_km0101" => $_POST["companycarmilage_km0101_$milage_year"], "companycarmilage_km3112" => $_POST["companycarmilage_km3112_$milage_year"]);
+    $_lib['db']->db_update_hash($post_values, 'companycarmilage', array("CompanyCarID" => $CompanyCarID, "MilageYear" => $milage_year));
+  }
+  $_lib['db']->db_update_hash($_POST, $db_table, $primarykey);
+}
+elseif($_lib['input']->getProperty('action_car_new'))
+{
+
+	if($_lib['input']->getProperty('companycar_CompanyCarID') > 0) {
```
Comment (sls-lnluksa) [inline modules/car/view/record.inc]: Redirect to car.list if there is no CarID.
Assign property as variable for later use.

File: modules/car/view/record.inc [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+		$query  = "select CompanyCarID from $db_table where CompanyCarID=" . (int) $_lib['input']->getProperty('companycar_CompanyCarID') . "";
+		$result = $_lib['db']->db_query($query);
+		if($_lib['db']->db_numrows($result) == 0)
+		{
+			$post = $_POST;
+			$post['companycar_ValidFrom'] = "NOW()";
+			$post['companycar_Active']  = 1;
+
+			$query  = "select CompanyCarID from $db_table where CompanyCarID=" . (int) $_lib['input']->getProperty('companycar_CompanyCarID') . "";
```
Comment (sls-lnluksa) [inline modules/car/view/record.inc]: Change this variable to insert query from line below.

File: modules/car/view/record.inc [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+			$_lib['db']->db_insert2(array('query'=>"insert into companycar (CompanyCarID, ValidFrom, Active) values (". (int) $post['companycar_CompanyCarID'].", ". $post['companycar_ValidFrom'] .", '". (int) $post['companycar_Active']."')"));
+			$CompanyCarID = $_POST['companycar_CompanyCarID'];
+		}
+		else
+		{
+			$CompanyCarID = $_POST['companycar_CompanyCarID'];
+		}
```
Comment (sls-lnluksa) [inline modules/car/view/record.inc]: Put `$CompanyCarID = $_POST['companycar_CompanyCarID'];` outside if block and remove else block.

File: modules/invoicein/view/edit.php [hunk 13]
```diff
@@ -224,6 +225,7 @@
         </td>
         <td><? print $_lib['form3']->text(array('table'=>$db_table2, 'field'=>'ProductNumber', 'pk'=>$LineID, 'value'=>$row2->ProductNumber, 'width' => 20, 'maxlength' => 20, 'tabindex'=>$tabindex++)) ?></td>
         <td><? print $_lib['form3']->text(array('table'=>$db_table2, 'field'=>'ProductName'  , 'pk'=>$LineID, 'value'=>$row2->ProductName,   'width' => 20, 'maxlength' => 80, 'tabindex'=>$tabindex++)) ?></td>
+        <td><? if($row2->CarID) { ?><? $_lib['form2']->car_menu2(array('table' => $db_table, 'field' => 'CarID', 'value' => $row2->CarID, 'tabindex' => $tabindex++)); } ?></td>
```
Comment (sls-lnluksa) [inline modules/invoicein/view/edit.php]: If reskonto has car enabled, than show it.

File: modules/report/view/reskontrovoucherprint.php [hunk 14]
```diff
@@ -193,6 +206,13 @@ function get_saldo($account, $fromperiod, $toperiod)
 
             // go trough account plans that are not in vouchers for the selected period
             // but still have some leftover amount from the time before starting peroid
+
+            // get first and last account id if they are not selected
+            if (!$_reskontroFrom && !$_reskontroTo) {
+              $acc_start_end  = $_lib['db']->get_row(array("query" => "SELECT MIN(AccountPlanID) AS MinAccPlanID, MAX(AccountPlanID) AS MaxAccPlanID FROM accountplan WHERE AccountPlanType = '$_type'"));
+              $_reskontroFrom = $acc_start_end->MinAccPlanID;
+              $_reskontroTo   = $acc_start_end->MinAccPlanID;
```
Comment (sls-lnluksa) [inline modules/report/view/reskontrovoucherprint.php]: `$_reskontroTo   = $acc_start_end->MinAccPlanID;` should be MaxAccPlanID

**Comment by stoivo** [comment]
Test out with @hildegunn95

**Comment by stoivo** [comment]
Remaining issues:
Car edir:
- [x] 1.new line for every 10th year
- [x] 2.add Kjøretøygruppe (Car type), dropdown,
- [x] 3.save button on left side
- [x] 4. if purchase date is "Simon" don't create YearMilage from year 0

Other
- [x] Add CarID here before Prosj http://puu.sh/l8lVW/f679604917.png

What is "Aktiver MVA"? just explain it so we know

Edit: 2. More info in next comment

**Comment by stoivo** [comment]
This is the new fields we want:

```
Registration number         Registreringsnummer
Car brand and modell        Merke og modell
Type of car             Kjøretøygruppe (dropdown)**
Registration year           Registreringsår
Amount of seats         Antall seter
Fuel                        Drivstoff

**
Varebil(klasse 2) 
Personbil
Lastebil(tankbil)
Slepvogn
```

**Comment by sls-lnluksa** [comment]
lodo.php?t=car.edit
Drivstoff should be select between DIESEL, BENSIN or ELEKTRISK
Slutt milage => Slutt km.stand

**Comment by hildegunn95** [comment]
- [x] Registreringsår only 4 digits (registration year)
- [x] Drivstoff is not displayed after it’s saved
- [x] Typen => Type
- [x] Bil navn => Bilnavn
- [x] Format number in km.stand
- [x] Round 'kr per kilometer' to two decimals
- [x] Saldo balance, 'Not selected car'(ikke valgt) option in dropdown for cars
- [x] Remove car from reskontro ‘saldobalanse’, ‘bilagsutskrift’ and ‘åpne poster’.
- [x] Select a car on a journal and save. Than make the car inactive. Refresh the journal edit and see. Now we can not see the car, something we should be able to. And if re-activate the car, the car appears in the journal again.

**Comment by stoivo** [comment]
- [x] All accounts between 7000 and 7099 should have car enabled
- [x] On car edit view reorder fields to: Active, Registration number, ... Number of seats, Car name, ...
- [x] Trim spaces from registration number on save (both on the edges and inside)
- [x] Voucher tax line should have the same car as it's parent line (update on change)
- [x] Realign on reports list page Reskontro > Saldobalanse: Avdeling, Prosjekt, Art should all be in the same line 
- [x] On voucher edit page if the car is selected and is not active it should be shown in the dropdown (but marked as inactive) otherwise list only active cars
- [x] Implement fetching car info from fakturabank (the unique link between fakturtabank and lodo should be registration number which needs to be without spaces)
- [x] Implement basic car info fetching on fakturabank (separate task)
- [x] Implement all these fields to fakturabank (create a separate task)

**Pull request merged**